be-typescript-express/files/README.md

```diff
@@ -0,0 +1,33 @@
+# be-typescript-express
+
+## Project setup
+
+### Configure .npmrc
```

Comment:
I don't know, if a section about the ``.npmrc`` file is actually needed. The quickstarter should work without this file. If this file is needed, e.g. for using non-public libraries, it would recommend to add it via ``Jenkinsfile`` and Jenkins Credentials plugin like this:
```
def stageBuild(def context) {
  stage('Build') {
    withCredentials([file(credentialsId: 'npmrc', variable: 'NPMRC')]) {
      sh "cp $NPMRC .npmrc"
      sh "npm install"
      ...
```
This would help to keep the nexus credentials out of the codebase.

Reply:
@cschweikert How would the dependencies be installed on the local machine then? Shouldn't they be pulled via Nexus as well?

Reply:
If you don't configure any registry (e.g. via ``.npmrc`` file), the dependencies are gathered from https://registry.npmjs.org directly. The dependencies inside the ``package.json`` of this quickstarter are available via https://registry.npmjs.org. Nexus usually works as a proxy to that registry.

Since we have a Nexus instance already in place, it would make sense to actually use it. AFAIK we are not setting or copying something (e.g. a ``.npmrc`` file) that would configure Nexus to be used automatically in all of our npm- or yarn-based builds.

If a developer wants to use this Nexus instance, configuring this via the ``.npmrc`` file locally is a good way to go. But a developer should NOT commit such a file, since it contains secrets. For this it would make sense to put it to the ``.gitignore``. Since it now only exists locally it also needs to be injected as a secret during a build process like mentioned above.

Reply:
Hmm, then it would make sense then to rename the header to `Configure .npmrc locally` instead and the comments about npmrc configs so that it addresses a wider audience.
```
Configure your .npmrc file to download the npm dependencies from your private remote repository.
```
As for the Jenkinsfile, I'll update it as well...

Reply:
@ManuelFeller If I got your PR #287 right, then the agent image actually ships with a global npmrc pointing to Nexus, right?

@rianet If I remember correctly, you had a different setup in our last project, right? Didn't you use env vars in the committed `.npmrc`?

I think it would be great to have one unified handling of `.npmrc` across all the quickstarters ... as far as I understand it our choices are either a) commit `.npmrc` WITHOUT credentials or b) ignore `.npmrc` in Git but help the user to set one up locally. Right?

Reply:
@michaelsauter correct, we now have the two node slaves with the matching `.npmrc` in the slave itself.

**But:** we either need to agree on all devs having this setting too on their dev machines (else the `package-lock.json` will point to the official registry **AND** need to be aware that this does not solve the issue for Ionic since they build on their slaves, so we need to keep it there.
So maybe we keep it for now to make sure nothing breaks when you are not aware of this (if you have the non-nexus url in the lockfile the slaves `.npmrc` will also not help - and that may lead to the proxy issues again).

_ToDo:_ documentation and alignment... ;-)

Reply:
Another note on the question how @rianet did it in her last project:
yes, they used env-vars since this way also works with Ionics build services.

It could be a good idea to have some guidance how to do so locally and have it prepared in all containers that builds node stuff...

Reply:
Then I think it makes sense to remove `.npmrc` config from README as we're not dependent on any internal libs like you said @cschweikert . @michaelsauter  If I understood you correctly, when you said you wanted a unified way of handling `.npmrc` across all quickstarters, you favor this one to also use `.npmrc`? Is this to make it future proof in case developers add libs from internal registry?

Reply:
It's not that easy.

I we remove it here we need to have it documented (and in our company communicate) that all users developing node based projects with ODS **must** use a local, system wide `.npmrc` pointing to our nexus. (Or have one not checked in to the projects.

The image for building node apps (node 10 and node 12) have this active, using the configured internal nexus. But if the developer does not have the same system wide `.npmrc` enabled for his system (and also no local `.npmrc` in the project) npm will install from the public registry and write these URL's of the packages into the `package.lock.json`. And that file overrides any registry url set elsewhere -> builds will then not use the nexus repository.

Also, we do not control the Ionic build environment, so there for sure we need to keep the `.npmrc` in the codebase.

What we could do is having a section in the readme that explains how to configure your local machine to use the nexus that is configured in the used ODS installation and teach how important that is before you execute a `npm install` (thus generating the `package-lock.json`)

As said before, I would not do any changes to the way how we treat `.npmrc` in this with PR but come up with a more complete solution regarding all Quickstarters that use node based tech. Because for sure we need to communicate / document where we differ from the everyday node development a new developer is most likely used to...

Reply:
Agreed ;)

Maybe we also add `.npmrc` into the `/be-typescript-express/files/.gitignore`. This will prevent any accidential commits of the `.npmrc` file.

Reply:
I am resolving this conversation. I think a separate issue on this would be highly beneficial to keep some quickstarters consistent across guidelines/documentation around `.npmrc`. We can always update this quickstarter once we reach a consensus there. 
Since libraries we install for the quickstarters are available in the official registry, changes regarding `.npmrc` shouldn't be a blocker against this PR.

Reply:
Correct, for our Ionic project we commit ```.npmrc``` with env vars, we use the secrets in ionic dashboard to build and package the app using nexus, not the public npm registry. But this is a very specific use case for Ionic and the usage of Ionic Dashboard.

For other npm based projects ignoring the file and raising the awareness to set it up locally using nexus or your own registry it's fine. However, what I like about the approach using env vars is that as a developer when you clone your repo and ```npm install``` you are reminded to set up your local env vars and the usage of nexus.